ext/gmp/gmp.c

```diff
@@ -1281,20 +1281,28 @@ ZEND_FUNCTION(gmp_fact)
 		RETURN_THROWS();
 	}
 
+#if SIZEOF_SIZE_T == 4
+	const zend_long maxbits = ULONG_MAX / GMP_NUMB_BITS;
+#else
+	const zend_long maxbits = INT_MAX;
+#endif
+
+
 	if (Z_TYPE_P(a_arg) == IS_LONG) {
-		if (Z_LVAL_P(a_arg) < 0) {
-			zend_argument_value_error(1, "must be greater than or equal to 0");
+		if (Z_LVAL_P(a_arg) < 0 || Z_LVAL_P(a_arg) > maxbits) {
+			zend_argument_value_error(1, "must be between 0 and " ZEND_LONG_FMT, maxbits);
 			RETURN_THROWS();
 		}
 	} else {
 		mpz_ptr gmpnum;
 		gmp_temp_t temp_a;
 
 		FETCH_GMP_ZVAL(gmpnum, a_arg, temp_a, 1);
+		long r = mpz_get_si(gmpnum);
```

Comment:
Please replace the call to `zval_get_long(a_arg)` below to use the new variable.

Reply:
Don't use `mpz_get_si()` here in the first place, since it may overflow for large `gmpnum`. Possibly use `mpz_sizeinbase(gmpnum, 2)` instead; that might need some fine-tuning of `maxbits`, though.